**Support Vector machine with different kernels(linear,gaussian,polynomial) and also tune the various parameters such as C ,gamma and degree to find out the best performing model.**

I am taking voice dataset of male and female

**Importing Libraries**

**Note:** here i am using %matplotlib inline, bcz the %matplotlib inline will make our plot outputs appear and be stored within the notebook.  

In [33]:
import pandas as pd
import numpy as np
import seaborn as sns

import matplotlib.pyplot as plt

%matplotlib inline

In [34]:
df = pd.read_csv('../input/voice.csv')
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...    centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...    0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...    0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...    0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...    0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...    0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

**correlation between each feature**

In [35]:
df.corr()

meanfreq        sd    median       Q25       Q75       IQR  \
meanfreq  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
sd       -0.739039  1.000000 -0.562603 -0.846931 -0.161076  0.874660   
median    0.925445 -0.562603  1.000000  0.774922  0.731849 -0.477352   
Q25       0.911416 -0.846931  0.774922  1.000000  0.477140 -0.874189   
Q75       0.740997 -0.161076  0.731849  0.477140  1.000000  0.009636   
IQR      -0.627605  0.874660 -0.477352 -0.874189  0.009636  1.000000   
skew     -0.322327  0.314597 -0.257407 -0.319475 -0.206339  0.249497   
kurt     -0.316036  0.346241 -0.243382 -0.350182 -0.148881  0.316185   
sp.ent   -0.601203  0.716620 -0.502005 -0.648126 -0.174905  0.640813   
sfm      -0.784332  0.838086 -0.661690 -0.766875 -0.378198  0.663601   
mode      0.687715 -0.529150  0.677433  0.591277  0.486857 -0.403764   
centroid  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
meanfun   0.460844 -0.466281  0.414909  0.545035  0.155091 -0.534462   
minfun    0.383937 -0.345609  0.337602  0.320994  0.258002 -0.222680   
maxfun    0.274004 -0.129662  0.251328  0.199841  0.285584 -0.069588   
meandom   0.536666 -0.482726  0.455943  0.467403  0.359181 -0.333362   
mindom    0.229261 -0.357667  0.191169  0.302255 -0.023750 -0.357037   
maxdom    0.519528 -0.482278  0.438919  0.459683  0.335114 -0.337877   
dfrange   0.515570 -0.475999  0.435621  0.454394  0.335648 -0.331563   
modindx  -0.216979  0.122660 -0.213298 -0.141377 -0.216475  0.041252   

              skew      kurt    sp.ent       sfm      mode  centroid  \
meanfreq -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
sd        0.314597  0.346241  0.716620  0.838086 -0.529150 -0.739039   
median   -0.257407 -0.243382 -0.502005 -0.661690  0.677433  0.925445   
Q25      -0.319475 -0.350182 -0.648126 -0.766875  0.591277  0.911416   
Q75      -0.206339 -0.148881 -0.174905 -0.378198  0.486857  0.740997   
IQR       0.249497  0.316185  0.640813  0.663601 -0.403764 -0.627605   
skew      1.000000  0.977020 -0.195459  0.079694 -0.434859 -0.322327   
kurt      0.977020  1.000000 -0.127644  0.109884 -0.406722 -0.316036   
sp.ent   -0.195459 -0.127644  1.000000  0.866411 -0.325298 -0.601203   
sfm       0.079694  0.109884  0.866411  1.000000 -0.485913 -0.784332   
mode     -0.434859 -0.406722 -0.325298 -0.485913  1.000000  0.687715   
centroid -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
meanfun  -0.167668 -0.194560 -0.513194 -0.421066  0.324771  0.460844   
minfun   -0.216954 -0.203201 -0.305826 -0.362100  0.385467  0.383937   
maxfun   -0.080861 -0.045667 -0.120738 -0.192369  0.172329  0.274004   
meandom  -0.336848 -0.303234 -0.293562 -0.428442  0.491479  0.536666   
mindom   -0.061608 -0.103313 -0.294869 -0.289593  0.198150  0.229261   
maxdom   -0.305651 -0.274500 -0.324253 -0.436649  0.477187  0.519528   
dfrange  -0.304640 -0.272729 -0.319054 -0.431580  0.473775  0.515570   
modindx  -0.169325 -0.205539  0.198074  0.211477 -0.182344 -0.216979   

           meanfun    minfun    maxfun   meandom    mindom    maxdom  \
meanfreq  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
sd       -0.466281 -0.345609 -0.129662 -0.482726 -0.357667 -0.482278   
median    0.414909  0.337602  0.251328  0.455943  0.191169  0.438919   
Q25       0.545035  0.320994  0.199841  0.467403  0.302255  0.459683   
Q75       0.155091  0.258002  0.285584  0.359181 -0.023750  0.335114   
IQR      -0.534462 -0.222680 -0.069588 -0.333362 -0.357037 -0.337877   
skew     -0.167668 -0.216954 -0.080861 -0.336848 -0.061608 -0.305651   
kurt     -0.194560 -0.203201 -0.045667 -0.303234 -0.103313 -0.274500   
sp.ent   -0.513194 -0.305826 -0.120738 -0.293562 -0.294869 -0.324253   
sfm      -0.421066 -0.362100 -0.192369 -0.428442 -0.289593 -0.436649   
mode      0.324771  0.385467  0.172329  0.491479  0.198150  0.477187   
centroid  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
meanfun   1.000000  0.339387  0.311950  0.270840

**Checking whether there is any null values**

In [36]:
df.isnull().sum()

meanfreq    0
sd          0
median      0
Q25         0
Q75         0
IQR         0
skew        0
kurt        0
sp.ent      0
sfm         0
mode        0
centroid    0
meanfun     0
minfun      0
maxfun      0
meandom     0
mindom      0
maxdom      0
dfrange     0
modindx     0
label       0
dtype: int64

In [37]:
df.shape

(3168, 21)

In [38]:
print("Total number of labels: ",df.shape[0])
print("Total number of male: ",df[df["label"] == "male"].shape[0])
print("Total number of male: ",df[df["label"] == "female"].shape[0])

Total number of labels:  3168
Total number of male:  1584
Total number of male:  1584


In [39]:
Y = df["label"]
df.drop(["label"], axis = 1, inplace=True)

In [40]:
X = df
X.shape

(3168, 20)

**Converting string value to int type for labels**

In [41]:
Y = Y.replace("male",1)
Y = Y.replace("female",0)
# we can also do it with scikit-learn using below code
# from sklearn.preprocessing import LabelEncoder
# y=df.iloc[:,-1]
# gender_encoder = LabelEncoder()
# y = gender_encoder.fit_transform(y)


**Data Standardisation**

Standardization refers to shifting the distribution of each attribute to have a mean of zero and a standard deviation of one (unit variance). It is useful to standardize attributes for a model. Standardization of datasets is a common requirement for many machine learning estimators implemented in scikit-learn; they might behave badly if the individual features do not more or less look like standard normally distributed data.


In [42]:
# Scale the data to be between -1 and 1
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)

**Spliting dataset into training and testing dataset**

In [43]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=1)

In [44]:
X_train.shape

(2534, 20)

**SVM with default hyperparameter.**

In [45]:
from sklearn.svm import SVC
SVC = SVC()
SVC.fit(X_train,y_train)
y_pred = SVC.predict(X_test)
from sklearn import metrics
print("Accuracy: ", metrics.accuracy_score(y_test,y_pred))

Accuracy:  0.9763406940063092


/opt/conda/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


**RBF kernel**

In [46]:
from sklearn.svm import SVC
SVC = SVC(kernel="rbf")
SVC.fit(X_train,y_train)
y_pred = SVC.predict(X_test)
from sklearn import metrics
print("Accuracy: ", metrics.accuracy_score(y_test,y_pred))

/opt/conda/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


Accuracy:  0.9763406940063092


**Ploynomial kernel**

In [47]:
from sklearn.svm import SVC
SVC = SVC(kernel="poly")
SVC.fit(X_train,y_train)
y_pred = SVC.predict(X_test)
from sklearn import metrics
print("Accuracy: ", metrics.accuracy_score(y_test,y_pred))

/opt/conda/lib/python3.6/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


Accuracy:  0.9589905362776026



Taking all the values of C and checking out the accuracy score with kernel as linear.

The C parameter tells the SVM optimization how much you want to avoid misclassifying each training example. For large values of C, the optimization will choose a smaller-margin hyperplane if that hyperplane does a better job of getting all the training points classified correctly. Conversely, a very small value of C will cause the optimizer to look for a larger-margin separating hyperplane, even if that hyperplane misclassifies more points.

Thus for a very large values we can cause overfitting of the model and for a very small value of C we can cause underfitting.Thus the value of C must be chosen in such a manner that it generalised the unseen data well



Taking kernel as rbf and taking different values gamma

Technically, the gamma parameter is the inverse of the standard deviation of the RBF kernel (Gaussian function), which is used as similarity measure between two points. Intuitively, a small gamma value define a Gaussian function with a large variance. In this case, two points can be considered similar even if are far from each other. In the other hand, a large gamma value means define a Gaussian function with a small variance and in this case, two points are considered similar just if they are close to each other


In [48]:
from sklearn.svm import SVC
c_range = list(range(1,30))
scores = {}
g_range = [0.1,0.01, 0.001, 0.0001]
for g in g_range:
    for  c in c_range:
        from sklearn.svm import SVC
        SVC = SVC(kernel="rbf",C=c,gamma=g)
        SVC.fit(X_train,y_train)
        y_pred = SVC.predict(X_test)
        from sklearn import metrics
        scores[metrics.accuracy_score(y_test,y_pred)] = (c,g)

In [49]:
print(scores.keys())

dict_keys([0.9779179810725552, 0.9810725552050473, 0.9794952681388013, 0.9763406940063092, 0.9747634069400631, 0.973186119873817, 0.9668769716088328, 0.9716088328075709, 0.9574132492113565, 0.9589905362776026, 0.9621451104100947, 0.9684542586750788, 0.9700315457413249, 0.8801261829652997, 0.8927444794952681, 0.9037854889589906, 0.9132492113564669, 0.919558359621451, 0.9290220820189274, 0.9400630914826499, 0.9479495268138801, 0.9526813880126183, 0.9542586750788643, 0.9558359621451105, 0.9605678233438486, 0.9637223974763407])


In [50]:
maximum_acc = (max(scores.keys()))
print("Maximum accuracy is:",maximum_acc)
print("values of c and g for maximum accuracy :",scores[maximum_acc])

Maximum accuracy is: 0.9810725552050473
values of c and g for maximum accuracy : (2, 0.1)


**So we achived maximum accuracy: 98.1% with c=2 and gamma=0.1**